File: app/src/main/java/com/rsam/customgrapher/unused/Filter.java
package com.rsam.customgrapher.unused;;

public class Filter {
	private static final int ORDER = 32;
	private static final boolean IIR = false;
    private float[] inputs = {0};
	private float[] outputs = {0};
	private float[] a = {0};
	private float[] b = {0};
	private float temp = 0;

    public void addVal(int value) {
		// Shifting data
        int i;
		for(i = ORDER-1; i > 0; i--) {
			inputs[i] = inputs[i - 1];
			if (IIR) outputs[i] = outputs[i - 1];
		}

		// Get the new first data, so current input
		inputs[0] = value;

		// Calculate the convolution from koefs
		temp = 0;
		for (i = 0; i < ORDER; i++) temp += inputs[i] * b[i];
		if (IIR) for (i = 1; i < ORDER; i++) temp -= outputs[i] * a[i];

		// Result
		outputs[0] = temp;
    }

	public float getVal() {
		return outputs[0];
	}
}

File: arduino/Oximeter/Filter.cpp
#include "Arduino.h"
#include "Filter.h"
using namespace std;

// Usage must addVal then getVal immediately since there's no buffer

// IIR if a is given
Filter::Filter(int _order, float *_b, float *_a) {
    // Using "_" to distinguish member and not and avoiding using "_".
    order = _order;
    a = _a;
    b = _b;
    
    int aSize = sizeof(a) / sizeof(float);
    int bSize = sizeof(b) / sizeof(float);
    
    if (aSize != 0) iir = true;
    init();
}

// FIR
Filter::Filter(int _order, float *_b) {
    order = _order;
    b = _b;
    iir = false;

    int bSize = sizeof(b) / sizeof(float);
    
    init();
}

void Filter::init() {
    // The last two brackets isn't necessary, just used to zero them
    if (iir) outputs = new float[1]();
    else outputs = new float[order]();
    inputs = new float[order]();
}

int st = 0; // The array index for x[n] in circular buffer

// Add value and simultanteously compute the result
void Filter::addVal(int value) {
    
    // Shifting the start of circular buffer FORWARD, since
    // the prev x(n) is x(n-1) now.
    st = modulo((st + 1), order);

    // Get input and store to the new start, x[n]
    inputs[st] = value;

    // Calculate the convolution from koefs
    outputs[st] = 0;
    
    for (int i = 0; i < order; i++) {
        int j = modulo((order + st - i), order); // Decrement, x[n-i] * b[i]
        
        outputs[st] += inputs[j] * b[i];
        
        if ((iir) && (i != 0)) {
            outputs[st] -= outputs[j] * a[i];
        }
    }
    
}

// Get individual output
float Filter::getVal() {
    return outputs[st];
}

// Alternate modulo function
// Good enough if dividend is sure to be not far bigger than divisor
int Filter::modulo(int dividend, int divisor) {
    if (dividend == 1) {
        return divisor > 1;
    } else if (divisor == 1) {
        return 0;
    } else if (divisor <= 4) {
        return dividend % divisor;
    } else {
        int result = dividend;
        for (result; result >= divisor; result -= divisor);
        return result;
    }
}
